### ISSUE TITLE
Add recount / reset earnings tools to core

### ISSUE THREAD

**Issue by pippinsplugins** [easydigitaldownloads/Easy-Digital-Downloads#3633]
Let's merge the [recount earnings extension](https://github.com/easydigitaldownloads/EDD-Recount-Earnings) into core and improve it.

- Use bulk processing
- Option to recount earnings / sales on products
- Option to reset earnings / sales on products (should NOT delete payments)
- Option to recount store-wide earnings (or maybe this is part of recounting products)

**Comment by cklosowski** [comment]
I'm going to pull down what you have and finish this up.

**Comment by pippinsplugins** [comment]
:+1:

**Comment by cklosowski** [comment]
Ok, so this is my initial work on it. Currently it'll do recount the store earnings, recount an individual download's earnings & sales, recount ALL downloads earnings & sales w/ store earnings update, and then a reset earnings and sales.

*NOTE:* Currently this _does not_ delete the sales logs, so the reset leaves those present. I'm doing this for the mean time, to allow the recounts to be much easier to test. I've still go to work out the performance side of deleting logs, etc.

**Comment by cklosowski** [comment]
Some UX for you:
![screen shot 2015-09-22 at 11 14 05 am](https://cloud.githubusercontent.com/assets/1390124/10027332/3946419a-611b-11e5-9dd7-d7d308e0e8eb.png)
![screen shot 2015-09-22 at 11 09 14 am](https://cloud.githubusercontent.com/assets/1390124/10027334/3965cc5e-611b-11e5-8e5d-913455fd0cef.png)
![screen shot 2015-09-22 at 11 09 08 am](https://cloud.githubusercontent.com/assets/1390124/10027336/39663d74-611b-11e5-94c6-bf37ac451e7b.png)
![screen shot 2015-09-22 at 11 14 08 am](https://cloud.githubusercontent.com/assets/1390124/10027337/3966a17e-611b-11e5-9dce-09e28e5717a3.png)
![screen shot 2015-09-22 at 11 14 16 am](https://cloud.githubusercontent.com/assets/1390124/10027333/3961e1a2-611b-11e5-94f3-acc121ad9bc4.png)
![screen shot 2015-09-18 at 4 15 05 pm](https://cloud.githubusercontent.com/assets/1390124/10027335/3966200a-611b-11e5-93fe-2041ca364d4f.png)

**Comment by cklosowski** [comment]
Easy compare link: https://github.com/easydigitaldownloads/Easy-Digital-Downloads/compare/release/2.5...issue/3633

**Comment by spencerfinnell** [comment]
😍 love it

**Comment by pippinsplugins** [comment]
@cklosowski The product drop down will have to use Chosen.

**Comment by cklosowski** [comment]
@pippinsplugins Ugh, fine. When can we set fire to all chosen things? It's a terrible UI.

**Comment by cklosowski** [comment]
The main issue is that Chosen does really poorly with elements show/hiding. I'll put a little hack together so it doesn't do this to us:
![screen shot 2015-09-22 at 12 59 16 pm](https://cloud.githubusercontent.com/assets/1390124/10029934/c51db7f8-6129-11e5-8a86-4ce5a86fb0f9.png)

**Comment by cklosowski** [comment]
Ok, the Reset option now fully deletes customers, payments, and any `edd_log` post types. I've added a filter on the post types and the deletion processes to allow other extensions to have their data deleted as well.

**Comment by cklosowski** [comment]
I need to work on building a UX for the 'confirmation' of full deletion. This seems like something we should make sure they want to do before we actually do it :)

**Comment by pippinsplugins** [comment]
Is there anything that's not working yet that we should be aware of while testing?

**Comment by cklosowski** [comment]
@pippinsplugins Not that i'm aware of. All of the options should be working just fine.

**Comment by cklosowski** [comment]
Confirmation UI
![screen shot 2015-09-24 at 11 48 55 am](https://cloud.githubusercontent.com/assets/1390124/10083213/4580afa2-62b2-11e5-8aa6-058c91004550.png)

**Comment by pippinsplugins** [comment]
@cklosowski I'd like to add another element to this: resetting customer stats. I just had a real case where a customer's stats were off by a lot (due to multiple customer records that didn't get merged properly). I had to reset them manually from the DB.

I'd like there to be two reset options:

1. Recount all customer stats in bulk
2. A button to recount a specific customer's stats. This should probably show on their details screen

**Comment by cklosowski** [comment]
Ok, reassigning this back to me to get the remainder of this done @pippinsplugins

**Comment by cklosowski** [comment]
@pippinsplugins do we want to 're-find' payments, or just recalculate the counts and lifetime value of the customer based off the payments in `$customer->payment_ids`?

**Comment by pippinsplugins** [comment]
Refind payments please.

**Comment by cklosowski** [comment]
I figured you were going to say that :/

**Comment by pippinsplugins** [comment]
Sorry, I retract that. Having thought about it a bit more, maybe we should only recount the ones in `payment_ids`

**Comment by cklosowski** [comment]
Yeah I think that's safest. Maybe once we get a merge tool built, and offer more front facing admin tools for customer management, we can move to re-finding their payments.

**Comment by pippinsplugins** [comment]
:+1:

**Comment by pippinsplugins** [comment]
The progress bar for `Recount Store Earnings` is jumping back and forth for me. One moment it's at 100%, the next moment it's at 30%, then it's at 100% again, then it's at 0%. It does that over and over.

**Comment by pippinsplugins** [comment]
My local host has a total of 29 sales but it's still going after several minutes.

**Comment by pippinsplugins** [comment]
It looks like it broke something too. I had to cancel the recount and once I did that, the dashboard widget now fails to load: https://cloudup.com/cqkPQyh2DYq - It loaded properly a moment before I ran the recount.

**Comment by pippinsplugins** [comment]
I ran recount store earnings a second time and it completed instantly. The dashboard widget remains broken though . .. super odd.

**Comment by pippinsplugins** [comment]
The recount for a single product worked properly on a product that had 0 sales. I then ran it on one that had 18 sales it is just sitting at 99% forever:
![screen shot 2015-10-22 at 6 24 28 pm](https://cloud.githubusercontent.com/assets/1034109/10680963/4182c590-78ea-11e5-8fd8-4621c1e1c9ea.png)

**Comment by pippinsplugins** [comment]
The `Recount All Store and Product Stats` is also sitting at 99% for me.

**Comment by cklosowski** [comment]
Ok thanks @pippinsplugins I'll look into those, sounds like i'm just missing something signaling the end of the report.

**Comment by SDavisMedia** [comment]
All of the selections seem to be working for me. I manually changed values in the database then used the tool to update them. All good. Two things, though.

1. **Reset Store:** I have no idea what it does. I have a general idea just from talking with @cklosowski about it some time ago but as of right now, I can't remember. The description tells me nothing, except that it's important. Anything we can do about that?

2. **Recount Customer Stats:** The submit button is grayed out and nothing else changed about the UI. I'm not sure what that one does.

**Comment by pippinsplugins** [comment]
1. About about adding a help text blurb that appears below the form for each option. When an option is selected, we can replace the blurb with the appropriate help text. This is an area we don't want to leave up to misinterpretation so the more guidance / explanation we can provide the better.

2. Not sure why it'd be grayed out but it recounts the `Total Purchase Count` and `Total Amount Spent` values for customer records.

**Comment by cklosowski** [comment]
@SDavisMedia Fixed teh thing about the grayed out buttons. Was just a UX logic thing when selecting different options.

@pippinsplugins I tried this again, none of my items hang, is there any errors reported in the responses of the AJAX requests that maybe i'm not seeing?

**Comment by pippinsplugins** [comment]
@cklosowski it may have been a one-off thing. I'll test again/

**Comment by cklosowski** [comment]
Going to work on adding some descriptions to these today.

**Comment by cklosowski** [comment]
Added in a customer specific recount tool, so you can just recount the sales and value of a specific customer by adding a 'Tools' menu item to the Customer UI:

![screen shot 2015-11-04 at 4 33 40 pm](https://cloud.githubusercontent.com/assets/1390124/10955787/960fdef8-8313-11e5-976a-e387755ce8fd.png)

**Comment by cklosowski** [comment]
Also added dismissible `X` to any of the batch processor success messages, as the discussion with @SDavisMedia went.

**Comment by pippinsplugins** [comment]
:+1:

**Comment by pippinsplugins** [comment]
A few UI thoughts here:

1. It would be awesome if we could avoid the "jump" caused by the addition of the help text when selecting an item.
![screen shot 2015-11-14 at 7 33 00 pm](https://cloud.githubusercontent.com/assets/1034109/11166691/a1bb738e-8b06-11e5-8b6e-0b4ee312eff3.png)
![screen shot 2015-11-14 at 7 33 36 pm](https://cloud.githubusercontent.com/assets/1034109/11166693/a52062aa-8b06-11e5-9d8e-68da24a92858.png)

2. The order of elements when resetting the store doesn't flow right. The "Are you sure?" prompt should be shown before the submit button, but right now it's down below.
![screen shot 2015-11-14 at 7 33 36 pm](https://cloud.githubusercontent.com/assets/1034109/11166697/c43660e0-8b06-11e5-8acc-40e9810481f3.png)

I love how it works, it just seems a bit odd.

Any thoughts on how that could be positioned better?

**Comment by cklosowski** [comment]
The jump isn't somethings I could avoid die to our wrappers

And the confirm being below, I discussed it with Sean and it's such an important act we don't want it to be too easy. It's a nuke button.

**Comment by pippinsplugins** [comment]
K, good enough for me.

**Comment by pippinsplugins** [comment]
Let's put the full reset at the complete bottom instead of having it above reset customer.

**Comment by pippinsplugins** [comment]
Customer recount appears to be missing a specific case.

I manually deleted a payment record from wp_posts and then recounted the customer stats. The purchase count did not change.

**Comment by pippinsplugins** [comment]
The logic looks right so I'm not seeing where it's missing the deleted record.

**Comment by pippinsplugins** [comment]
Going to that single customer's detail screen and recounting sats for just that customer also misses the deleted payment.

**Comment by cklosowski** [comment]
Deleting from wp_posts wouldn't delete the proper logs and might still count as a payment.

**Comment by cklosowski** [comment]
Could still be seeing post meta for the payment?

**Comment by pippinsplugins** [comment]
@cklosowski That was my thought but that doesn't make sense to me as the recount purchases is just using the `post__in` parameter for the `EDD_Payment_Query`, so that should be coming from `wp_posts` where `post_type=edd_payment`, not the logs.

**Comment by cklosowski** [comment]
Not sure. Will have to test. Does it work correctly of you delete a payment from the payments list table instead of directly in the DB?

**Comment by pippinsplugins** [comment]
@cklosowski I couldn't test properly doing that as that method would have triggered our deletion routine and the stats would have been updated. I wanted to test in non-standard ways to ensure we can account for when data has gone missing due to unknown reasons.

**Comment by cklosowski** [comment]
I've been testing it my manual changing the value and the payment count of the customer.

**Comment by cklosowski** [comment]
Missing data that way proves to be a greater issue with our payment stats class it may seem.

**Comment by cklosowski** [comment]
@pippinsplugins Give that a shot, I changed the order in which we set the purchase value to 0, and also check if the payment is NULL before trying to process it.

**Comment by pippinsplugins** [comment]
Testing now.

I just noticed a separate problem though.

This customer has one completed purchase of $10, yet the recount made the customer's value `0.00`:
![screen shot 2015-11-15 at 8 52 39 am](https://cloud.githubusercontent.com/assets/1034109/11169192/415cd882-8b76-11e5-960b-1afde57c564e.png)

**Comment by pippinsplugins** [comment]
Ok rerunning the recount did the following:

1. Sale count is still at `2` even though it should be `1`
2. The recount never finishes. It just keeps restarting
3. The `Lifetime Value` is flipping between `$10` and `$20` as the recount restarts over and over

**Comment by cklosowski** [comment]
Did you pull the latest?

**Comment by pippinsplugins** [comment]
Yes.

**Comment by cklosowski** [comment]
The early decision to just use the payments recorded in the payment ids column is what we're running up against. If you manually delete it from the DB it's not removed from their list. That's what we'd need to resolve with a storage of all payment ids that retun null on get post and then on the last step remove those from the payment ids column and then recount.

No clue what's up wth your value part. I can't break this deleting from the DB directly after that last update.

**Comment by pippinsplugins** [comment]
@cklosowski I don't see why that's a reason for this to fail. The point of this tool is precisely to solve the problem of when records have been edited, deleted, or skewed in an unintended manner.

**Comment by cklosowski** [comment]
Not saying it's a reason to release it broken, saying that's the route we're going got have to go down, correcting that lost data in the customer's table, which wasn't something we fully discussed in the beginning. The tool as it stands in this PR is far more in depth than our github repo tool ever was, so that's going to make it so much more useful upon release.

**Comment by pippinsplugins** [comment]
I don't think I'm following, simply because the issue I'm seeing is one of the exact reasons this tool was being built into core, at least from my memory :)

**Comment by cklosowski** [comment]
Ok. Let me look through some of the logic. Does it work for someone with multiple payments? Just trying to narrow down the possible cause.

**Comment by cklosowski** [comment]
Ok my single payment one isn't doing this. @pippinsplugins let me know when you are around so we can try and debug this together.

**Comment by pippinsplugins** [comment]
To do:
- [ ] Reset Sequential order numbers
- [ ] Remove scroll bar: https://cloudup.com/cMAjFXANmhF

**Comment by pippinsplugins** [comment]
This has now been merged into `release/2.5` for string lock.

**Comment by mindctrl** [comment]
"Please" is misspelled in the dropdown

![image](https://cloud.githubusercontent.com/assets/1254870/11253138/bc69c68c-8e07-11e5-80cf-93f49a37828f.png)

**Comment by mindctrl** [comment]
Question about these options:
- "Recount Store Earnings and Sales"
- "Recount Earnings and Sales for All downloads"

Do these options not do the same thing?

**Comment by cklosowski** [comment]
@pippinsplugins @mindctrl Also, this issue has some additions to the WP CLI class that allows you to send in an email fname and lname into the payment creation command, so you can create payments with different details.

**Comment by cklosowski** [comment]
Tested this with a download with more than 1000 sales, a customer with over 1000 sales, and just general thousands. The EDD Site staging will be a great test.

**Comment by pippinsplugins** [comment]
@cklosowski Something is still wrong with customer recount. I just did a site wide customer recount and results are not as expected for one particular customer.

Many, many payments for this customer:

Details screen said there were just 2 sales for the customer.

I recounted, now it shows only 1: https://cloudup.com/c5uYCqzO295

It should be 19.

**Comment by pippinsplugins** [comment]
It's due to the recount logic:
```php
foreach ( $customers as $customer ) {

	$payment_ids    = explode( ',', $customer->payment_ids );

	$payment_args = array(
		'post__in' => $payment_ids,
		'number'   => -1,
	);

	$payments = edd_get_payments( $payment_args );

	$purchase_value = 0.00;
	$payment_ids    = array();

	foreach ( $payments as $payment ) {
		$purchase_value += edd_get_payment_amount( $payment->ID );
		$payment_ids[] = $payment->ID;
	}

	$purchase_count = count( $payment_ids );
	$payment_ids    = implode( ',', $payment_ids );

	$customer_update_data = array(
		'purchase_count' => $purchase_count,
		'purchase_value' => $purchase_value,
		'payment_ids'    => $payment_ids,
	);

	$customer_instance = new EDD_Customer( $customer->id );
	$customer_instance->update( $customer_update_data );

}
```

That only looks for payments already attached to the customer. This really needs to also find all payments *NOT* attached to the customer properly.

Instead of querying just payment IDs attached to the customer, how about, instead, we do a meta query to find payments that match the customer email.

**Comment by cklosowski** [comment]
See above. We discussed this. So we want it changed now?

**Comment by pippinsplugins** [comment]
Yes.

**Comment by cklosowski** [comment]
:+1:

**Comment by mindctrl** [comment]
Tested a different site and see something similar to https://github.com/easydigitaldownloads/Easy-Digital-Downloads/issues/3633#issuecomment-157863855.

Created a bunch of payments on a new site with wp-cli. Then I marked some payments as refunded and some as pending. I get this:
![image](https://cloud.githubusercontent.com/assets/1254870/11547671/c00d1ace-9923-11e5-9de5-17cccd0db570.png)

I ran Recount Earnings and Sales for All Downloads and get this:
![image](https://cloud.githubusercontent.com/assets/1254870/11547702/fe91283a-9923-11e5-8f68-d871864e888c.png)

I ran Recount Store Earnings and Sales and get this:
![image](https://cloud.githubusercontent.com/assets/1254870/11547716/1d71f31a-9924-11e5-91f6-0c9c96f007b9.png)

Note the Total Earnings number changes. I can keep running those two options and it goes back and forth to the same numbers every time.

**Comment by cklosowski** [comment]
Ok thanks @mindctrl probably just a status change in my get_paymetns request.

**Comment by cklosowski** [comment]
@mindctrl ready for another go please?

**Comment by pippinsplugins** [comment]
@cklosowski `Recount customer stats` gave all customers the same values:
![screen shot 2015-12-02 at 10 50 04 pm](https://cloud.githubusercontent.com/assets/1034109/11551423/13415cfa-9947-11e5-848e-b4746def6f4e.png)

**Comment by pippinsplugins** [comment]
@mindctrl Do you get the same behavior I do?

**Comment by pippinsplugins** [comment]
Ok now it should work. @mindctrl could you test?

**Comment by mindctrl** [comment]
I'm still seeing a weird difference in `Total Earnings` on the dashboard when running those two tools. Here's how to reproduce it.

- Use wp-cli to create some payments. I'm doing 500. Here are my initial stats:
![image](https://cloud.githubusercontent.com/assets/1254870/11562984/4ae36434-999f-11e5-9915-a74aa9fcb4f0.png)

- Then mark some payments pending. I did the first 30 on the screen.
- Run `Recount Earnings and Sales for All Downloads`. Here are my updated stats:
![image](https://cloud.githubusercontent.com/assets/1254870/11563014/758c304e-999f-11e5-86a7-1ca27d31f50d.png)

- Run `Recount Store Earnings and Sales`. My updated stats:
![image](https://cloud.githubusercontent.com/assets/1254870/11563042/a6bbecae-999f-11e5-885f-6693f43b4179.png)

I've done this 3 times now, and I get similar results. Something in the `Recount Earnings and Sales for All Downloads` is not right.

**Comment by mindctrl** [comment]
Wiped out. Created 500 new payments, split equally between 2 customers. 

![image](https://cloud.githubusercontent.com/assets/1254870/11563451/f8d9270c-99a1-11e5-8add-bb747b33c50a.png)

Ran `Recount Customer Stats`:
![image](https://cloud.githubusercontent.com/assets/1254870/11563459/0360d9e0-99a2-11e5-9a51-76976d706f8f.png)

**Comment by mindctrl** [comment]
Added another 250 payments for a new customer
![image](https://cloud.githubusercontent.com/assets/1254870/11563540/7d7ac3f8-99a2-11e5-8a02-e135a2bd5125.png)

Ran customer recount again:
![image](https://cloud.githubusercontent.com/assets/1254870/11563544/85b2f34c-99a2-11e5-8ae4-af7c8d6006bb.png)

**Comment by pippinsplugins** [comment]
It looks like that all customers are getting the right stats except for the last customer who gets everything.

**Comment by mindctrl** [comment]
Each of those should have 250 payments. It's doubling for every customer I add, and then recount.

![image](https://cloud.githubusercontent.com/assets/1254870/11563741/7d912fc0-99a3-11e5-872a-c50ef8a09a1f.png)

**Comment by mindctrl** [comment]
After the 'latest' I should say.

**Comment by mindctrl** [comment]
Customer count issue fixed. The Recount for All Downloads is still weird. New site, 500 payments:

![image](https://cloud.githubusercontent.com/assets/1254870/11564288/7585c32e-99a6-11e5-9b08-7094d22463e6.png)

Then set latest 30 payments to pending and recount all downloads:
![image](https://cloud.githubusercontent.com/assets/1254870/11564323/9c746fd0-99a6-11e5-985a-ba0e4863819d.png)

**Comment by pippinsplugins** [comment]
@mindctrl does recount for individual downloads work properly?

**Comment by mindctrl** [comment]
No, it does something similar. 143 sales went down to 12 on the product I tested. Should be 131, because I marked 12 pending.

**Comment by pippinsplugins** [comment]
@mindctrl Could you test the old recount earnings plugin and recount the stats for one of the downloads and see if it is correct? https://github.com/easydigitaldownloads/EDD-Recount-Earnings

**Comment by mindctrl** [comment]
It changed the count to 143. It counted the pending payments.

**Comment by pippinsplugins** [comment]
@mindctrl Try recount all download stats again.

**Comment by mindctrl** [comment]
We narrowed this down in Slack to https://github.com/easydigitaldownloads/Easy-Digital-Downloads/issues/4052

**Comment by pippinsplugins** [comment]
Merged into `release/2.5` and closed until we find bugs.

**Comment by mindctrl** [comment]
This doesn't reset customer recurring subscription status.

**Comment by cklosowski** [comment]
This doesn't touch anything recurring. That should be done in the recurring extension.

**Comment by mindctrl** [comment]
So a recurring reset tool?

**Comment by pippinsplugins** [comment]
Yeah, recurring will need to hook into this to add another option.

@cklosowski Could you ensure that the dropdown is filterable?

**Comment by cklosowski** [comment]
No there are hooks for all the tools that the extensions cam get into.

**Comment by pippinsplugins** [comment]
@cklosowski This drop down is not filterable:
```
<select name="edd-export-class" id="recount-stats-type">
	<option value="0" selected="selected" disabled="disabled"><?php _e( 'Please select an option', 'easy-digital-downloads' ); ?></option>
	<option data-type="recount-store" value="EDD_Tools_Recount_Store_Earnings"><?php _e( 'Recount Store Earnings and Sales', 'easy-digital-downloads' ); ?></option>
	<option data-type="recount-download" value="EDD_Tools_Recount_Download_Stats"><?php printf( __( 'Recount Earnings and Sales for a %s', 'easy-digital-downloads' ), edd_get_label_singular( true ) ); ?></option>
	<option data-type="recount-all" value="EDD_Tools_Recount_All_Stats"><?php printf( __( 'Recount Earnings and Sales for All %s', 'easy-digital-downloads' ), edd_get_label_plural( true ) ); ?></option>
	<option data-type="recount-customer-stats" value="EDD_Tools_Recount_Customer_Stats"><?php _e( 'Recount Customer Stats', 'easy-digital-downloads' ); ?></option>
	<option data-type="reset-stats" value="EDD_Tools_Reset_Stats"><?php _e( 'Reset Store', 'easy-digital-downloads' ); ?></option>
</select>
```
I'd like that to be filterable so we can register new options in the dropdown.

**Comment by pippinsplugins** [comment]
We need to also remove the old recount earnings tab:
![screen shot 2015-12-08 at 10 52 34 am](https://cloud.githubusercontent.com/assets/1034109/11661931/e102ec02-9d99-11e5-90f2-2db82c42632a.png)
I vote we just call `remove_action()` to ensure the plugin doesn't load any more.

**Comment by pippinsplugins** [comment]
Just ran a recount on edd staging and it looks to have worked well.

**Comment by cklosowski** [comment]
@pippinsplugins how's that for filters?

**Comment by pippinsplugins** [comment]
That'll do.

**Comment by pippinsplugins** [comment]
I'm closing this :)

**Comment by pippinsplugins** [comment]
Something broke here. I just ran the recount all product stats tool live on pippinsplugins.com

It appears that only one product was recounted and it failed massively.

Before: https://cloudup.com/cTmxT28BVlJ
After: https://cloudup.com/c_UbJLiuSVo

**Comment by pippinsplugins** [comment]
I then ran the recount for just that product and it appears to have worked properly.

**Comment by pippinsplugins** [comment]
The recount all stats also failed on the total store sales / earnings: https://cloudup.com/cqJu9FO7qXI

**Comment by pippinsplugins** [comment]
I think it's due to the heavy reliance on update_option() in a heavily cached system.

**Comment by pippinsplugins** [comment]
Local works fine.

**Comment by pippinsplugins** [comment]
I think the best way to fix this will be to remove the use of `update_option()` and instead make direct DB queries.

**Comment by cklosowski** [comment]
:+1: makes sense.

**Comment by pippinsplugins** [comment]
@cklosowski is this ready to test again on live pippinsplugins.com?

**Comment by pippinsplugins** [comment]
@cklosowski I just tested it again and it's failing in exactly the same way on pippinsplugins.com: https://cloudup.com/cRZLgsMB9L1

**Comment by pippinsplugins** [comment]
When reseting the store on a local site that has about 4,000 payment records, I'm getting a memory error:

Fatal error: Allowed memory size of 134217728 bytes exhausted (tried to allocate 524288 bytes) in /wp-includes/wp-db.php on line 2344

Think we can do anything about that @cklosowski ?

**Comment by pippinsplugins** [comment]
The queries currently delete by ID:
```
switch( $type ) {
	case 'customers':
		$table_name = $wpdb->prefix . 'edd_customers';
		$sql[] = "DELETE FROM $table_name WHERE id IN ($ids)";
		break;
	case 'downloads':
		$sql[] = "UPDATE $wpdb->postmeta SET meta_value = 0 WHERE meta_key = '_edd_download_sales' AND post_id IN ($ids)";
		$sql[] = "UPDATE $wpdb->postmeta SET meta_value = 0.00 WHERE meta_key = '_edd_download_earnings' AND post_id IN ($ids)";
		break;
	case 'other':
		$sql[] = "DELETE FROM $wpdb->posts WHERE id IN ($ids)";
		$sql[] = "DELETE FROM $wpdb->postmeta WHERE post_id IN ($ids)";
		$sql[] = "DELETE FROM $wpdb->comments WHERE comment_post_ID IN ($ids)";
		$sql[] = "DELETE FROM $wpdb->commentmeta WHERE comment_id NOT IN (SELECT comment_ID FROM $wpdb->comments)";
		break;
}
```

We really should change that do just a `DELETE WHERE` and remove the `IN`.

**Comment by pippinsplugins** [comment]
The other way to fix it would be to delete 50 at a time instead of querying all records at once, as this does:

```
			$args = apply_filters( 'edd_tools_reset_stats_total_args', array(
				'post_type'      => $edd_types_for_reset,
				'post_status'    => 'any',
				'posts_per_page' => -1,
			) );

			$posts = get_posts( $args );
			foreach ( $posts as $post ) {
				$items[] = array(
					'id'   => (int) $post->ID,
					'type' => $post->post_type,
				);
			}

			$customer_args = array( 'number' => -1 );
			$customers     = EDD()->customers->get_customers( $customer_args );
			foreach ( $customers as $customer ) {
				$items[] = array(
					'id'   => (int) $customer->id,
					'type' => 'customer',
				);
			}
```

**Comment by cklosowski** [comment]
@pippinsplugins doing an `IN` will trigger 1 reindex. Going through and doing a DELETE statement for every `ID` we need to will trigger multiple reindexes...unless I'm misreading what exactly your `DELETE WHERE` statement is you are thinking of.

But by doing a `DELETE WHERE IN()` we're kicking off a single query, instead of one for every.

As far as that total, you can't query all 50 and then get a % change, that initial `pre_fetch` is to get the total number of items to delete, it's a GET and doesn't cause a reindex...by using this we can determine a complete %, and not have to query it on every step, since it's stored for the entire length of the process...

**Comment by pippinsplugins** [comment]
I was thinking of deleting all records with a single query. Problem with an `IN` statement is it can fail 100% if there are too many records.

**Comment by pippinsplugins** [comment]
Finally tracked it down!
